### PR TITLE
RS-19: Implement comprehensive scanner filtering with streaming commit processing

### PR DESCRIPTION
**Pull request by deeprave**
## Summary
- Implement full repository history scanning with streaming approach for memory efficiency
- Add AuthorPatternMatcher with glob wildcard support for flexible author filtering
- Enable git reference starting points and comprehensive query parameter integration

### Core Functionality
- Replace HEAD-only traversal with full repository history scanning
- Add streaming approach for memory-efficient processing of large repositories
- Implement proper commit count statistics instead of hardcoded values

### Author Filtering Enhancement
- Add AuthorPatternMatcher with glob pattern support for wildcard matching
- Support consolidated filtering: email patterns (contains '@') vs name patterns
- Implement case-insensitive pattern matching for improved usability
- Add auto-completion for incomplete email patterns (@domain → *@domain)
- Support both include and exclude filtering with exclude precedence

### Git Reference Support
- Enable traversal starting from any git reference (branch, tag, commit SHA)
- Add proper reference resolution with annotated tag handling
- Default to HEAD when no git_ref specified for backward compatibility

### Query Parameter Integration
- Implement max_commits limiting during commit traversal
- Add date range filtering support for time-based queries
- Update scanner to use new QueryParams field structure from RS-15

### Technical Improvements
- Refactor scanner ID generation with configurable hash length (12 chars)
- Update scan execution to use streaming message processing
- Add proper error handling for git reference resolution failures
- Implement memory-efficient commit processing with callback pattern

## Test Plan
- [x] All 26 scanner tests passing
- [x] Added comprehensive test coverage for all filtering scenarios
- [x] Verified wildcard patterns, case sensitivity, and edge cases
- [x] Added 650+ lines of test code covering complex filtering combinations
- [x] Memory-efficient streaming tested with large repositories

## Related Issues
- Implements RS-19: Update scanner to handle QueryParams field changes from RS-15
- Supports RS-15: CLI changes for scanner filters and repository specification

🤖 Generated with [Claude Code](https://claude.ai/code)

### CODE REVIEW

File: src/core/pattern_parser.rs [hunk 1]
```diff
@@ -172,6 +172,112 @@ pub fn parse_extensions(extensions_str: &str) -> Vec<String> {
         .collect()
 }
```
[the remark this answers is not in the document]
Comment (deeprave) [reply src/core/pattern_parser.rs:173]: This would have undesirable side-effects, such as requiring the dot '.' character in the domain.
